File: debop4s-core/src/test/java/debop4s/core/jtests/LocalTest.java
package debop4s.core.jtests;

import debop4s.core.AbstractValueObject;
import debop4s.core.Local;
import debop4s.core.testing.Testing;
import debop4s.core.utils.Hashs;
import debop4s.core.utils.ToStringHelper;
import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.UUID;
import java.util.concurrent.Callable;

import static org.fest.assertions.Assertions.assertThat;

public class LocalTest {

    private static final Logger log = LoggerFactory.getLogger(LocalTest.class);

    @Before
    public void before() { Local.clearAll(); }

    @Test
    public void multiThread() throws Exception {
        Testing.run(15, new Runnable() {
            @Override
            public void run() {
                try {
                    saveAndLoadValueType();
                    saveAndLoadReferenceType();
                    getOrCreate();
                } catch (Exception e) {
                    log.error("예외 발생", e);
                }
            }
        });
    }

    @Test
    public void saveAndLoadValueType() {
        final String key = "Local.Value.Key.Java";
        final String value = UUID.randomUUID().toString();
        Local.put(key, value);
        assertThat(Local.get(key).getOrElse(null)).isEqualTo(value);
    }

    @Test
    public void saveAndLoadReferenceType() throws Exception {
        final String key = "Local.Reference.Key.Java";
        final User user = new User("user", "P" + Thread.currentThread().getId(), 1);
        Local.put(key, user);

        Thread.sleep(5);

        User storedUser = Local.get(key).getOrElse(null);

        assertThat(storedUser).isNotNull();
        assertThat(storedUser).isEqualTo(user);
        assertThat(storedUser.name).isEqualTo(user.name);
    }

    @Test
    public void getOrCreate() throws Exception {
        String key = "Local.GetOrCreate.Key.Java";
        User user = Local.getOrCreate(key, new Callable<User>() {
            @Override
            public User call() throws Exception {
                return new User("user", "P" + Thread.currentThread().getId(), 1);
            }
        }).getOrElse(null);

        Thread.sleep(5);

        User storedUser = Local.get(key).getOrElse(null);
        assertThat(storedUser).isNotNull();
        assertThat(storedUser).isEqualTo(user);
        assertThat(storedUser.name).isEqualTo(user.name);
    }

    static class User extends AbstractValueObject {
        private String name;
        private String password;
        private int age;

        public User(String name, String password, int age) {
            this.name = name;
            this.password = password;
            this.age = age;
        }

        @Override
        public int hashCode() {
            return Hashs.compute(name, password, age);
        }

        @Override
        public ToStringHelper buildStringHelper() {
            return super.buildStringHelper()
                        .add("name", name)
                        .add("password", password)
                        .add("age", age);
        }

        private static final long serialVersionUID = 2697543433170138506L;
    }
}

File: debop4s-core/src/test/scala/debop4s/core/stests/LocalTest.scala
package debop4s.core.stests

import debop4s.core._
import java.util.UUID
import org.slf4j.LoggerFactory

/**
 * debop4s.core.tests.LocalTest
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since  2013. 12. 14. 오후 7:11
 */
class LocalTest extends AbstractCoreTest {

  override lazy val log = LoggerFactory.getLogger(getClass)

  before { Local.clearAll() }

  test("put/get value type") {
    val key = "Local.Value.Key"
    val value = UUID.randomUUID().toString
    Local.put(key, value)

    if (key.isWhitespace) fail("key is whitespace")
    val stored = Local.get(key).getOrElse(null).asInstanceOf[String]
    assert(stored != null)
    assert(stored == value)
  }

  test("put/get reference type") {
    val key = "Local.Reference.Key"
    val user = new User("user", "P" + Thread.currentThread().getId, 1)
    Local.put(key, user)
    Thread.sleep(5)

    val storedUser = Local.get[User](key).getOrElse(null)
    assert(storedUser != null)
    assert(storedUser === user)
    assert(user.name === storedUser.name)
    assert(user.password === storedUser.password)
    assert(user.age === storedUser.age)
  }

  test("get or create if not exists") {
    val key = "Local.GetOrCreate.Key"
    val user = Local.getOrCreate(key, {
      new User("user", "P" + Thread.currentThread().getId, 2)
    }).getOrElse(null)

    Thread.sleep(5)

    val storedUser = Local.get[User](key).getOrElse(null)
    assert(storedUser != null)
    assert(storedUser === user)
    assert(storedUser.name === user.name)
    assert(storedUser.password === user.password)
    assert(storedUser.age === user.age)
  }
}

case class User(name: String, password: String, age: Int)